### ISSUE TITLE
Localforage / Cordova Issue: removeItem and clear not working in cordova-based  hybrid mobile apps.

### ISSUE THREAD

**Issue by vijeysrc** [localForage/localForage#179]
I also posted this issue in stackoverfow: http://stackoverflow.com/questions/24115938/localforage-cordova-issue-removeitem-and-clear-not-working

We use localforage for data storage in our cordova-based hybrid mobile application. As a part of a cleanup operation, we used localforage.clear() to remove all the stored data. The data is not getting cleaned up. We see this issue both in devices and ios simulator. In ios simulator, we debugged this using the safari develop tools. After the cleanup operation, we still see the data in "Resources > Databases".

In another part of the application, we tried to remove a single item using removeItem(). That also did not work.

Therefore both clear() and removeItem() method of localforage is not working in the hybrid mobile application.

The interesting point is: When we ran this application in our desktop, the same clean up operations are working fine.

NOte: We use indexedDB as the storage technology behind the localforage.

**Comment by oleduc** [comment]
Same when using https://github.com/ocombe/angular-localForage and PhantomJS.

**Comment by tofumatt** [comment]
Are you explicitly forcing IndexedDB? localForage shouldn’t use it in Cordova, but will on Chrome/Firefox. I would have thought Cordova was always using WebSQL.

Could you send a test case to try in cordova/desktop/etc. so I can reproduce the issue?

**Comment by skddc** [comment]
We just ran into this as well. iOS/Cordova.

**Comment by stevemadere** [comment]
I have hit this same problem with Cordova on Android.
Cordova v 5 or 6, Android V 5.1.

In my case I am using localForage to store a queue of images to upload
in a background job so it is of critical importance.  After queueing about 12 MB
of images (and calling $localForage.removeItem() on each of them
after each finishes uploading), all subsequent calls to $localForage.setItem() fail.
The calls to removeItem() all supposedly succeeded since they invoked
the success callback.

The OS app manager shows that the app is now using
12 MB of data space even though I have called removeItem() on every single
thing that I had previously called setItem on.  I have to delete all of my Cordova
app's data via the app manager to get back to a state where $localForage.setItem()
will succeed again.

Along with cleaning up the leakage from the queue thrashing, I also lose the cached
authorization token and any other locally stored personalization and option settings.
Fun.

**Comment by guopengliang** [comment]
We used localForage v1.5.0 with the localforage.INDEXEDDB driver on iOS 10.3, and experience the dramatic increase in system storage usage. Reports from users, some increased a few hundred MB's and some even up to 13GB, while their actual records size is <=2MB!

We are still not sure what causes such increase, or what is in it. Now we have switched to the localforage.WEBSQL driver (still on localForage v1.5.0) alongside with the previously stored INDEXEDDB database. Surprisingly, even when we no longer read, write or initiates the INDEXEDDB database, system storage did not go down. 

If this is what @stevemadere said a leakage from queue trashing, is there any way to safely eliminate it without completely reset the app? 

![viewfile](https://cloud.githubusercontent.com/assets/188754/24584154/2d04e4b2-1796-11e7-9068-add82c9f153a.gif)

**Comment by thgreasi** [comment]
Have you tried to create a new localforage instance using IndexedDB and then use the `clear()` method?
You may need to use the actual IndexedDB API to drop the IndexedDB database completely.
So, it seems that iOS IndexedDB implementation still has issues?
@tofumatt How about releasing v1.5.1/v1.6 to use WebSQL on iOS and keep iOS IndexedDB under a flag?

**Comment by guopengliang** [comment]
Yes, I have tried that. But it didn't help, the storage usage didn't go down.

**Comment by guopengliang** [comment]
@thgreasi I have tried to delete the entire IndexedDB with `var req = indexedDB.deleteDatabase('localforage');`, but somehow the request was blocked (`req.onblock` event called) and the delete operation failed.

Perhaps this is because the IndexedDB itself was trying to execute the previously failed commands in the queue trashing, and also causing the trash to pile up taking that much of space? Is there any other way I could force delete it?

**Comment by thgreasi** [comment]
@guopengliang I am not aware of an other way, sorry.

**Comment by Pittan** [comment]
I'm experiencing to similar issue. 
Incrementing DB version number every time when `createInstance` will  
clear storage space.  
But the storage still grows up until re-creating instance just like before.

**Comment by thgreasi** [comment]
@Pittan On what LF version, driver, browser do you experience that?


 when createInstance will clear storage space.
> But the storage still grows up until re-creating instance just like before.

This was confusing for me, can you provide more info?
Specifically
> Incrementing DB version number every time

Are you doing this yourself?


>  when createInstance will clear storage space.

CreateInstance doens't do something like this on its own. Do you mean that you use createInstance and then clear from the returned LF instance?

> grows up until re-creating instance just like before.

Some more info or a pseudocode would get me back on track.

**Comment by Pittan** [comment]
@thgreasi 
I'm using LF v1.5.3 with iOS 11.1.2 & 11.2.1 using from AngularJS 1.5.
[The document](https://localforage.github.io/localForage/#settings-api-config) says the version parameter of config is not used, but it seems to be working.  

So, when creating instance of localforage, I wrote like this

```typescript
const ver = new Date().getTime() / 100
console.log(`open IndexedDB: ${ver}`)
this.localforage = $window.localforage.createInstance({
  driver: $window.localforage.INDEXEDDB,
  name:   'imageCache',
  version: ver
})
```
It will solve this issue temporally, but it will cause `ConstraintError`.

In my iOS App, writing a lot of data to IndexedDB will make a huge `.wal` file to its storage.

Also, when I wrote another code to use IndexedDB directly, I found that closing DB expressly after writing will clean up the `.wal` file.
Maybe something like `flush()` will solve this issue.

More later, I will make a repro code (if I can 😄 )

**Comment by thgreasi** [comment]
>  I found that closing DB expressly

So you mean that doing `this.localforage._dbInfo.db.close()` cleans up your `.wal` file?

**Comment by Pittan** [comment]
No, I was saying about raw IndexedDB APIs. 
But I tested `this.localforage._dbInfo.db.close()` and it works, too!

**Comment by Pittan** [comment]
I created a [test repository](https://github.com/Pittan/localforage-test) and a [test page](https://pittan.github.io/localforage-test/).
When the page is loaded, it will fetch an image and store it to indexedDB using localforage.  

-- 
When image data is already in DB, it will update the date.
Press "Reload" to test that behavior. 
Press "Reset" to close db connection.

**Comment by thgreasi** [comment]
When v1.6 gets released we will be confident that the idb driver will be able to resume from most db states. We could have a connection timeout to close the connection after some time of inactivity, but for performance reasons we currently want the connection to be alive all the time.

**Comment by thgreasi** [comment]
Also, the latest v1.5 release should be used if you decide to call db.close() from within your app.

**Comment by thgreasi** [comment]
Is this still an issue?
I saw at https://github.com/localForage/localForage/issues/811#issuecomment-436447728 that a related webkit issue was fixed.
See: https://bugs.webkit.org/show_bug.cgi?id=191294#c13
